nexus/src/db/lookup.rs

```diff
@@ -0,0 +1,723 @@
+    use futures::future::BoxFuture;
+    use omicron_common::api::external::LookupResult;
+
+    // TODO-dap XXX-dap we could probably get rid of this trait and its impls
+    // because we're now calling `lookup_root()` from a macro that can
+    // essentially inline the impl of GetLookupRoot for Key.
+    pub trait GetLookupRoot {
+        fn lookup_root(&self) -> &LookupPath<'_>;
+    }
+
+    pub trait LookupNoauthz {
+        type LookupType;
+        fn lookup(&self) -> BoxFuture<'_, LookupResult<Self::LookupType>>;
+    }
+}
+
+use private::GetLookupRoot;
+use private::LookupNoauthz;
+
+pub trait LookupFor {
+    type LookupType;
+    fn lookup_for(
+        &self,
+        action: authz::Action,
+    ) -> BoxFuture<'_, LookupResult<Self::LookupType>>;
+}
+
+impl<T> LookupFor for T
+where
+    T: LookupNoauthz + GetLookupRoot + Send + Sync,
+    <T as LookupNoauthz>::LookupType:
+        AuthorizedResource + Clone + std::fmt::Debug + Send,
+{
+    type LookupType = <T as LookupNoauthz>::LookupType;
+    fn lookup_for(
+        &self,
+        action: authz::Action,
+    ) -> BoxFuture<'_, LookupResult<Self::LookupType>> {
+        async move {
+            let lookup = self.lookup_root();
+            let opctx = &lookup.opctx;
+            let rv = self.lookup().await?;
+            opctx.authorize(action, &rv).await?;
+            Ok(rv)
+        }
+        .boxed()
+    }
+}
+
+enum Key<'a, P> {
+    Name(P, &'a Name),
+    Id(LookupPath<'a>, Uuid),
+}
+
+impl<'a, T> GetLookupRoot for Key<'a, T>
+where
+    T: GetLookupRoot,
+{
+    fn lookup_root(&self) -> &LookupPath<'_> {
+        match self {
+            Key::Name(parent, _) => parent.lookup_root(),
+            Key::Id(lookup, _) => lookup,
+        }
+    }
+}
+
+impl<'a, P> Key<'a, P> {
+    fn lookup_type(&self) -> LookupType {
+        match self {
+            Key::Name(_, name) => LookupType::ByName(name.as_str().to_string()),
+            Key::Id(_, id) => LookupType::ById(*id),
+        }
+    }
+}
+
+pub struct LookupPath<'a> {
+    opctx: &'a OpContext,
+    datastore: &'a DataStore,
+}
+
+impl<'a> LookupPath<'a> {
+    pub fn new<'b, 'c>(
+        opctx: &'b OpContext,
+        datastore: &'c DataStore,
+    ) -> LookupPath<'a>
+    where
+        'b: 'a,
+        'c: 'a,
+    {
+        LookupPath { opctx, datastore }
+    }
+
+    pub fn organization_name<'b, 'c>(self, name: &'b Name) -> Organization<'c>
+    where
+        'a: 'c,
+        'b: 'c,
+    {
+        Organization { key: Key::Name(self, name) }
+    }
+
+    pub fn organization_id(self, id: Uuid) -> Organization<'a> {
+        Organization { key: Key::Id(self, id) }
+    }
+
+    pub fn project_id(self, id: Uuid) -> Project<'a> {
+        Project { key: Key::Id(self, id) }
+    }
+
+    pub fn instance_id(self, id: Uuid) -> Instance<'a> {
+        Instance { key: Key::Id(self, id) }
+    }
+
+    pub fn disk_id(self, id: Uuid) -> Disk<'a> {
+        Disk { key: Key::Id(self, id) }
+    }
```

Comment:
So to clarify, if we extend this model to other types, we need to:

- Add functions to `LookupPath` for each resource that can be looked up (generally "by ID")
- Add a `define_lookup_with_parent` macro call
- Add manual impls to look up names, for the auto-generated structs?

Reply:
With the new proc macro implementation, the steps are:

- add one function to `LookupPath` -- `fn ${new_resource_name}_id()`
- add a call to the new `lookup_resource` macro to define everything else for the new resource